### ISSUE TITLE
Popup windows should be centered on the screen

### ISSUE THREAD

**Issue by Timidger** [way-cooler/way-cooler#218]
Currently, we just float the windows and don't do anything with their positioning. However, by default this puts them in the top left corner (0, 0). Instead, they should be in the exact center of the screen.

* Grab the resolution of the screen
* Find the mid point
* Subtract by width and height of the view
* ???
* Profit

**Comment by platipo** [comment]
I'm experincing a strange behaviour with firefox download popup windows, they spawn in the top left corner

**Comment by Timidger** [comment]
Thanks, I'll look into trying to fix that. 

In the mean time, can you reproduce that in sway? If not, they are probably handling the logic correctly and I can base my fix on how they handle it.

**Comment by platipo** [comment]
Yes, I can reproduce the same faulty behaviour in sway.
Is there anything I can do to send it to you?

**Comment by Timidger** [comment]
No, I'll test it on my own machine so no need to send me anything. It's potentially a bug/limitation in either wlc or xwayland/Firefox if sway hasn't fixed it, but I'll dig into it and report back.

**Comment by Timidger** [comment]
@platipo Hmm, actually I can't seem to reproduce on my setup (using either the latest master version OR v0.6.0). What is your Firefox version? I'm using 54.0.1 so perhaps this is a bug that has later been fixed.

Why don't you test it and send me a simple log output from Way Cooler? Before you run it however, please enable these options:

`export WLC_DEBUG=render,xwm`

That will enable some WLC debugging that can help me figure out what the problem is.

**Comment by platipo** [comment]
Save version of Firefox, here it goes the log saved with
`way/cooler &>log.txt`
There is only the firefox part but still over 2000 lines
[log.txt](https://github.com/way-cooler/way-cooler/files/1151229/log.txt)

**Comment by Timidger** [comment]
Ok, looking over the logs highlights something interesting.

I see you're testing this by opening the download link at [the firefox download page](https://www.mozilla.org/en-US/firefox/new/?scene=2). I went there and tested that on my machine and it actually doesn't always spawn in the top left corner, sometimes it spawns in the center (though offset a bit, which I will have to look into fixing). This seems to tell me that it is a weird data race going on somewhere _probably_ between wlc/Wayland and Firefox. 

What makes me think this is these relevant lines from the logs:

```
 TRACE [callbacks] src/callbacks.rs:99 View WlcView { handle: 3, title: "Opening firefox-54.0.1.tar.bz2", class: "Firefox" } requested geometry Geometry { origin: Point { x: 920, y: 520 }, size: Size { w: 1, h: 1 } } 

...

 TRACE [callbacks] src/callbacks.rs:99 View WlcView { handle: 3, title: "Opening firefox-54.0.1.tar.bz2", class: "Firefox" } requested geometry Geometry { origin: Point { x: 0, y: 0 }, size: Size { w: 477, h: 338 } } 

... (that is repeated a few more times)
```

*Sometimes* this happens, sometimes it doesn't This might have to do with the size of the window actually. Note that in the first request the size is 1x1, far too small to be seen. I have code that makes sure it is the minimum size that a container can be when a window is spawned, but that will come as a *second request* later on after we set the new origin point. I'm thinking that Firefox is seeing that and wants the window to be the correct size but it's (obviously) not setting the origin point correctly.

I'll test this more and see if I can combine the two requests to see if this fixes the problem.

**Comment by Timidger** [comment]
Ok, I have pushed a branch that "fixes" the problem by not allowing dialog popups to be at origin point 0,0. This doesn't break any apps *that I know about*, perhaps some apps want to spawn a popup in the top left corner and this will break it. If a user moves it to 0,0 it should still work properly (and it's a bit hard to make that pixel perfect, so it probably won't come up). I have marked this appropriately as a hack that I can hopefully fix later.

Ultimately this is a tricky bug to fix properly. The best way in the future would be to wait for some time for all requests from clients to come through and then see if they made the geometry coordinates what we want (and if we don't like it, change it). That would require some sort of listener thread or timer that would then change the geometry as necessary once we think that all the requests have come through (e.g, after 100 ms or something). That still isn't _correct_, the app should just tell us where it wants to go and we should be able to put it there when the surface is spawned but that's part of the fun with working with XWayland 😢.

I'll test that branch over the next couple of days (and you are welcome to as well) to see if I caused any regressions. If not, it'll be delivered along with 0.6.1.

**Comment by platipo** [comment]
Problem is persistent in `v1:0.6.1-2`, it comes out opening virt-manager, popup windows spawn in (0, 0).
[250 lines log](https://github.com/way-cooler/way-cooler/files/1190018/log.txt)

**Comment by Timidger** [comment]
Will take a look this weekend, won't have time to until then

**Comment by Timidger** [comment]
Ok tried out `virt-manager` there is indeed issues. Have you noticed issues in any other applications?

I assume by your report that it worked correctly in `0.6.0`? E.g this is a regression in `0.6.1`?

**Comment by platipo** [comment]
Sadly haven't tried in `0.6.0`

**Comment by Timidger** [comment]
I tested on v0.6.0: The popup is still in the top left (😢), but the drop down menus are not messed up. I will fix the regression, and see if I can also find out why its spawning in the top left.

**Comment by Timidger** [comment]
@platipo  Just pushed a branch w/ a PR that fixes it (#358). If you notice any other issues in that pr feel free to report them. If you confirm that fixes it, I'll merge it in but hold off on updating to v0.6.2 until I have time to test that there are no more regressions (and obviously if any bugs are filed in that time I'll fix them so that they can be included in v0.6.2).
